flink-runtime/src/main/java/org/apache/flink/runtime/security/SecurityUtils.java

```diff
@@ -71,163 +64,93 @@
-							config.hadoopConf);
-						Method addCredentialsMethod = UserGroupInformation.class.getMethod("addCredentials",
-							Credentials.class);
-						addCredentialsMethod.invoke(loginUser, cred);
-					} catch (NoSuchMethodException e) {
-						LOG.warn("Could not find method implementations in the shaded jar. Exception: {}", e);
-					}
-				}
-			} else {
-				// login with current user credentials (e.g. ticket cache)
-				try {
-					//Use reflection API to get the login user object
-					//UserGroupInformation.loginUserFromSubject(null);
-					Method loginUserFromSubjectMethod = UserGroupInformation.class.getMethod("loginUserFromSubject", Subject.class);
-					Subject subject = null;
-					loginUserFromSubjectMethod.invoke(null, subject);
-				} catch (NoSuchMethodException e) {
-					LOG.warn("Could not find method implementations in the shaded jar. Exception: {}", e);
-				}
-
-				// note that the stored tokens are read automatically
-				loginUser = UserGroupInformation.getLoginUser();
+		// install the security modules
+		List<SecurityModule> modules = new ArrayList();
+		try {
+			for (Class<? extends SecurityModule> moduleClass : config.getSecurityModules()) {
+				SecurityModule module = moduleClass.newInstance();
+				module.install(config);
+				modules.add(module);
 			}
+		}
+		catch(Exception ex) {
+			throw new Exception("unable to establish the security context", ex);
+		}
+		installedModules = modules;
 
-			LOG.info("Hadoop user set to {}", loginUser.toString());
+		// install a security context
+		// use the Hadoop login user as the subject of the installed security context
+		if (!(installedContext instanceof NoOpSecurityContext)) {
+			LOG.warn("overriding previous security context");
+		}
+		UserGroupInformation loginUser = UserGroupInformation.getLoginUser();
+		installedContext = new HadoopSecurityContext(loginUser);
+	}
 
-			boolean delegationToken = false;
-			final Text HDFS_DELEGATION_KIND = new Text("HDFS_DELEGATION_TOKEN");
-			Collection<Token<? extends TokenIdentifier>> usrTok = loginUser.getTokens();
-			for (Token<? extends TokenIdentifier> token : usrTok) {
-				final Text id = new Text(token.getIdentifier());
-				LOG.debug("Found user token " + id + " with " + token);
-				if (token.getKind().equals(HDFS_DELEGATION_KIND)) {
-					delegationToken = true;
+	static void uninstall() {
+		if(installedModules != null) {
+			for (SecurityModule module : Lists.reverse(installedModules)) {
+				try {
+					module.uninstall();
 				}
-			}
-
-			if (!loginUser.hasKerberosCredentials()) {
-				//throw an error in non-yarn deployment if kerberos cache is not available
-				if (!delegationToken) {
-					LOG.error("Hadoop Security is enabled but current login user does not have Kerberos Credentials");
-					throw new RuntimeException("Hadoop Security is enabled but current login user does not have Kerberos Credentials");
+				catch(UnsupportedOperationException e) {
```

Comment:
The uninstall isn't used in production, only in test code.   Throwing gives more information to the unit test.  Hopefully getting rid of the warning will suffice for now.